app/client/components/payment/update/card/stripeCardInputForm.tsx

```diff
@@ -11,143 +11,43 @@ import { Spinner } from "../../../spinner";
-                display: "flex",
-                flexWrap: "wrap",
-                justifyContent: "space-between"
-              }}
-            >
-              {this.state.isValidating ? (
-                <>
-                  <StripeLogo />
-                  <Spinner
-                    loadingMessage="Validating your card details..."
-                    scale={0.7}
-                    inline
-                  />
-                </>
-              ) : (
-                <NavigateFnContext.Consumer>
-                  {nav =>
-                    nav.navigate ? (
-                      <>
-                        <a
-                          href="https://stripe.com/"
-                          target="_blank"
-                          css={{ marginBottom: "5px" }}
-                        >
-                          <StripeLogo />
-                        </a>
-                        <div
-                          css={{
-                            textAlign: "right",
-                            [maxWidth.mobileLandscape]: {
-                              width: "100%"
-                            }
-                          }}
-                        >
-                          <Button
-                            disabled={
-                              !(
-                                this.props.stripe &&
-                                this.props.stripeSetupIntent
-                              )
-                            }
-                            text="Review payment update"
-                            onClick={this.startCardUpdate(nav.navigate)}
-                            primary
-                            right
-                          />
-                        </div>
-                        {this.renderError()}
-                      </>
-                    ) : (
-                      <GenericErrorScreen loggingMessage="No navigate function - very odd" />
-                    )
-                  }
-                </NavigateFnContext.Consumer>
-              )}
-            </div>
-          </div>
-        </div>
-      </>
-    );
-  }
+export const StripeCardInputForm = (props: StripeCardInputFormProps) => {
+  const [isValidating, setIsValidating] = useState<boolean>(false);
+  const [readyElements, setReadyElements] = useState<string[]>([]);
+  const [error, setError] = useState<StripeInputFormError>({});
+  const stripe = useStripe();
+  const elements = useElements();
 
-  private isLoaded = () =>
-    this.props.stripe &&
-    this.state.readyElements.length === 3 &&
-    this.props.stripeSetupIntent;
+  const isLoaded = () => {
+    return stripe && readyElements.length === 3 && props.stripeSetupIntent;
+  };
 
-  private markElementReady = (element: string) => () =>
-    this.setState(prevState => ({
-      readyElements: prevState.readyElements.concat(element)
-    }));
+  const markElementReady = (element: string) => () =>
+    setReadyElements(readyElements.concat(element));
```

Comment:
this could use spread these days

Reply:
in fact, I suspect this the cause of the hanging on ![image](https://user-images.githubusercontent.com/19289579/86350685-f911e700-bc5a-11ea-81a1-71d06a3acf7e.png) as previously it used the overload of setState which was a function from old state to new, whereas this new hooks approach is just using the state available on the last render, which if these elements mark themselves as ready in quick succession between renders, then we'd get stuck

Reply:
as discussed IRL, maybe better to pull the readiness out into three little pieces of state to avoid this whole fiddlyness and code will be more readable too 🙂

Reply:
referenced this problem here https://github.com/guardian/manage-frontend/wiki/Major-TODOs#replace-all-react-class-based-components-with-hooks-based-functional-components as a gotcha should the team ever decide to migrate any/all the class based components to hooks